### ISSUE TITLE
WeeklyPaper 6. CNN layer, AutoEncoder(Encoder, Decoder)

### ISSUE THREAD

**Issue by finneKIM** [finneKIM/WeeklyPaper#7]
<html>
<body>
<h1>1. CNN을 구성하는 각 레이어의 역할</h1>
<h2>🧠 CNN을 구성하는 주요 레이어와 그 역할</h2>
<hr>
<h3>1. <strong>입력층 (Input Layer)</strong></h3>
<ul>
<li>
<p>🔹 <strong>역할</strong>: 이미지 데이터를 CNN에 넣어주는 출발점</p>
</li>
<li>
<p>🔹 예시: 가로 28, 세로 28, 흑백 이미지 → (1, 28, 28)</p>
<p>→ 1은 채널 수 (흑백이므로 1, 컬러는 보통 3)</p>
</li>
</ul>
<hr>
<h3>2. <strong>합성곱층 (Convolutional Layer)</strong></h3>
<ul>
<li>🔹 <strong>역할</strong>: 이미지에서 **특징(Feature)**을 추출하는 핵심 레이어</li>
<li>🔹 <strong>작동 방식</strong>:
<ul>
<li>필터(또는 커널)라는 작은 창을 이용해 이미지 전체를 스캔하면서,</li>
<li>각 위치에 맞춰 곱셈과 덧셈 연산을 진행해 <strong>특징 맵(Feature Map)</strong> 생성</li>
</ul>
</li>
<li>🔹 <strong>필터</strong>는 경계선, 모서리, 색상 차이 등을 감지하는 역할</li>
<li>🔹 여러 개의 필터 사용 → 다양한 특징을 동시에 추출 가능</li>
</ul>
<pre><code class="language-python">
nn.Conv2d(in_channels=1, out_channels=32, kernel_size=3, stride=1, padding=1)
# 1채널 입력 이미지에 대해 32개의 3x3 필터 적용

</code></pre>
<hr>
<h3>3. <strong>활성화 함수층 (Activation Layer - ReLU)</strong></h3>
<ul>
<li>🔹 <strong>역할</strong>: 비선형성을 부여하여 <strong>복잡한 패턴</strong>도 학습 가능하게 만듦</li>
<li>🔹 대표적인 함수: ReLU (Rectified Linear Unit)</li>
<li>🔹 동작: 0보다 작으면 0, 0보다 크면 그대로 출력</li>
<li>🔹 장점: 연산 간단 + Gradient 소실 방지</li>
</ul>
<pre><code class="language-python">F.relu(x)

</code></pre>
<hr>
<h3>4. <strong>풀링층 (Pooling Layer)</strong></h3>
<ul>
<li>🔹 <strong>역할</strong>: 이미지 크기를 줄여 <strong>계산량 감소</strong>, <strong>과적합 방지</strong></li>
<li>🔹 주로 사용되는 방식: <strong>Max Pooling</strong>
<ul>
<li>일정 영역에서 가장 큰 값만 가져옴</li>
</ul>
</li>
<li>🔹 장점:
<ul>
<li>이미지의 <strong>중요한 특징은 유지</strong>, 불필요한 정보는 줄임</li>
<li>모델을 더 <strong>일반화</strong>할 수 있게 해줌</li>
</ul>
</li>
</ul>
<pre><code class="language-python">nn.MaxPool2d(kernel_size=2, stride=2)
# 2x2 영역에서 가장 큰 값만 뽑아내고, 겹치지 않도록 stride 2

</code></pre>
<hr>
<h3>5. <strong>드롭아웃층 (Dropout Layer)</strong></h3>
<ul>
<li>🔹 <strong>역할</strong>: 학습 중 일부 뉴런을 임의로 제거해 <strong>과적합 방지</strong></li>
<li>🔹 작동 방식: 매 학습 단계마다 일부 노드를 &quot;끄기&quot;</li>
<li>🔹 효과: 학습 데이터에 너무 특화되는 걸 방지하여 <strong>일반화 성능</strong> 향상</li>
</ul>
<pre><code class="language-python">
nn.Dropout(p=0.5)
# 50% 확률로 뉴런을 비활성화

</code></pre>
<hr>
<h3>6. <strong>완전연결층 (Fully Connected Layer, FC)</strong></h3>
<ul>
<li>🔹 <strong>역할</strong>: 추출된 특징을 기반으로 최종 분류나 회귀 작업 수행</li>
<li>🔹 CNN의 뒷부분에서 <strong>클래스별 확률</strong> 등을 계산</li>
<li>🔹 일반적인 신경망(Dense Layer)과 유사</li>
</ul>
<pre><code class="language-python">
nn.Linear(in_features=512, out_features=10)
# 512개의 입력 특징 → 10개의 클래스 점수 출력

</code></pre>
<hr>
<h3>7. <strong>출력층 (Output Layer)</strong></h3>
<ul>
<li>🔹 <strong>역할</strong>: 최종 예측값을 확률로 변환</li>
<li>🔹 일반적으로 <strong>Softmax 함수</strong> 사용
<ul>
<li>모든 클래스의 값을 확률처럼 바꿔줌 (총합이 1이 되도록 정규화)</li>
</ul>
</li>
</ul>
<pre><code class="language-python">
F.softmax(x, dim=1)
# 클래스별 확률 출력

</code></pre>
<hr>
<h3>🔄 CNN 전체 구조 요약 (흐름도 순서)</h3>
<pre><code>
입력층 → [합성곱 → ReLU → 풀링] x N → 전개(Flatten) → FC층 → Softmax → 예측값 출력

</code></pre>
<hr>
<h3>📌 Flatten이란?</h3>
<ul>
<li>🔹 CNN의 마지막 <code>Conv + Pooling</code> 층까지 거친 결과는 2차원 형태(채널, 가로, 세로)</li>
<li>🔹 이를 1차원으로 <strong>펼쳐주는 것</strong>이 <code>Flatten</code></li>
<li>🔹 이후 Fully Connected Layer에 연결할 수 있음</li>
</ul>
<pre><code class="language-python">
x = x.view(-1, 512)  # 예: (batch_size, 512)로 reshape

</code></pre>
<h2>CNN 구성 요소 도식 &amp; 전체 코드 흐름도</h2>
<h2>📊 1. CNN 기본 구조 도식 (구성 요소 시각화)</h2>
<pre><code>
이미지 입력 (Input Layer)
       │
       ▼
 ┌───────────────┐
 │ 합성곱층 (Conv) │ ← 특징 추출
 └───────────────┘
       │
       ▼
 ┌──────────────┐
 │ 활성화 함수 (ReLU) │ ← 비선형성 부여
 └──────────────┘
       │
       ▼
 ┌──────────────┐
 │ 풀링층 (Pooling) │ ← 정보 요약/축소
 └──────────────┘
       │
      (반복)
       │
       ▼
 ┌────────────┐
 │ Flatten (전개) │ ← 2D → 1D
 └────────────┘
       │
       ▼
 ┌────────────────────┐
 │ 완전연결층 (Fully Connected) │ ← 분류 준비
 └────────────────────┘
       │
       ▼
 ┌────────────┐
 │ 출력층 (Softmax) │ ← 예측 확률 계산
 └────────────┘
       │
       ▼
  클래스 예측 결과

</code></pre>
<hr>
<h2>💻 2. CNN 전체 코드 흐름 예시 (PyTorch 기반)</h2>
<pre><code class="language-python">
import torch
import torch.nn as nn
import torch.nn.functional as F

# CNN 모델 정의
class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()

        # 1. 합성곱층 + ReLU + 풀링
        self.conv1 = nn.Conv2d(in_channels=1, out_channels=32, kernel_size=3, padding=1)
        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)

        # 2. 두 번째 합성곱층
        self.conv2 = nn.Conv2d(in_channels=32, out_channels=64, kernel_size=3, padding=1)

        # 3. 완전연결층
        self.fc1 = nn.Linear(in_features=64*7*7, out_features=128)  # 28x28 → 7x7까지 줄어듦
        self.fc2 = nn.Linear(in_features=128, out_features=10)      # 클래스 수: 10

        # 4. 드롭아웃
        self.dropout = nn.Dropout(0.5)

    def forward(self, x):
        # 입력: (batch_size, 1, 28, 28)
        x = self.pool(F.relu(self.conv1(x)))  # conv1 + ReLU + pooling
        x = self.pool(F.relu(self.conv2(x)))  # conv2 + ReLU + pooling

        x = x.view(-1, 64*7*7)  # Flatten
        x = self.dropout(F.relu(self.fc1(x)))  # FC1 + ReLU + Dropout
        x = self.fc2(x)                        # FC2 (출력층 전 단계)
        x = F.softmax(x, dim=1)                # 클래스별 확률
        return x

</code></pre>
<hr>
<h2>📌 참고 시각 자료 요약</h2>

레이어 종류 | 역할 요약 | 주요 기능 | 예시 코드
-- | -- | -- | --
Input Layer | 이미지 입력 | 데이터 형태: (채널, 높이, 너비) | -
Convolutional Layer | 특징 추출 | 필터/커널을 통해 특성 맵 생성 | nn.Conv2d()
ReLU Layer | 비선형 활성화 | 음수 제거, 연산 간소화 | F.relu()
Pooling Layer | 크기 축소, 특징 요약 | 주로 MaxPooling 사용 | nn.MaxPool2d()
Dropout Layer | 과적합 방지 | 일부 뉴런 무작위 제거 | nn.Dropout()
Flatten Layer | 2D → 1D 전개 | FC 연결 전 1차원 변환 | x.view()
Fully Connected Layer | 예측/분류 | 출력값 계산 | nn.Linear()
Output Layer | 확률 변환 | 클래스별 확률 출력 | F.softmax()

<hr>
<ul>
<li>출처
<ul>
<li><a href="https://yjjo.tistory.com/8">https://yjjo.tistory.com/8</a></li>
<li><a href="https://pyimagesearch.com/2021/05/14/convolutional-neural-networks-cnns-and-layer-types/">https://pyimagesearch.com/2021/05/14/convolutional-neural-networks-cnns-and-layer-types/</a></li>
</ul>
</li>
</ul>

</body>
</html>

**Comment by finneKIM** [comment]
# 2. 오코 인코더가 적용되기 적합한 상황

### 🟨 1) 차원 축소 (Dimensionality Reduction)

- 많은 특성을 가진 데이터를 **요약**하고 싶을 때 사용
- PCA(주성분 분석)의 비선형 버전이라고 생각하면 됨
- 예: 고해상도 이미지 → 저차원 벡터 → 다시 이미지 복원

---

### 🟨 2) 잡음 제거 (Denoising)

- **노이즈가 섞인 이미지**나 데이터에서 원래 내용을 복원
- 예: 스캔한 문서에 얼룩이 있거나, 손상된 사진을 복원하는 데 사용

---

### 🟨 3) 이상 탐지 (Anomaly Detection)

- 정상 데이터를 학습한 후, **복원 오류가 큰 데이터 = 이상치로 판단**
- 예: 정상적인 기계 소리 학습 → 이상한 소리 입력 시 복원이 잘 안됨 → 이상 감지

---

### 🟨 4) 이미지 생성, 데이터 생성 (Generative Modeling)

- VAE(변분 오토인코더), GAN과 결합하여 새로운 데이터 생성
- 예: 손글씨 숫자처럼 패턴이 있는 이미지를 자동 생성

---

### 🟨 5) 전처리 및 임베딩 벡터 생성

- **원본 데이터를 압축된 벡터로 표현** → 다른 모델 입력으로 사용 가능
- 예: 문장/이미지 → 벡터 임베딩 → 분류/추천 등에 활용

---

- 출처
    - https://velog.io/@gyurili/%EB%94%A5%EB%9F%AC%EB%8B%9D-%EC%98%A4%ED%86%A0%EC%9D%B8%EC%BD%94%EB%8D%94Autoencoder-%EC%A0%95%EB%A6%AC-%EC%9D%B8%EC%BD%94%EB%8D%94%EC%99%80-%EB%94%94%EC%BD%94%EB%8D%94
    - https://ssung-22.tistory.com/55
    - https://junstar92.tistory.com/113

**Comment by finneKIM** [comment]
<html>
<body>
<h1>2-1. 오토 인코더를 구성하는 encoder, decoder의 각각의 개념과 차이점</h1>
<p>오토인코더는 크게 2부분으로 구성됨:</p>
<pre><code>
입력 데이터 → [Encoder] → 잠재 표현(latent vector) → [Decoder] → 복원된 데이터

</code></pre>
<hr>
<h3>🔷 Encoder (인코더)</h3>

항목 | 설명
-- | --
정의 | 입력 데이터를 압축해 중요한 정보만 담은 **잠재 벡터(latent vector)**를 만드는 역할
목적 | 원본 데이터를 요약해 핵심 정보만 추출 (축소)
구성 | 여러 층의 CNN, LSTM, 또는 Dense Layer 사용 가능
결과 | 잠재 공간(latent space)의 벡터 z (입력보다 작음)


<h3>Autoencoder 구조</h3>
<pre><code>
        입력 데이터 (예: 손글씨 이미지)
                      │
                      ▼
       ┌────────────────────────┐
       │       Encoder          │  ← 특징 추출 및 압축
       │ (Conv → ReLU → Pool)   │
       └────────────────────────┘
                      │
                      ▼
        잠재 표현 (Latent Vector, z) ← 압축된 핵심 정보
                      │
                      ▼
       ┌────────────────────────┐
       │       Decoder          │  ← 복원
       │ (Upsample → Conv → ReLU)│
       └────────────────────────┘
                      │
                      ▼
            복원된 데이터 (예: 원본 이미지와 유사)

</code></pre>

</body>
</html>
